stock_production_lot_warranty/models/stock_production_lot.py

```diff
@@ -15,34 +15,25 @@ class StockProductionLot(models.Model):
 
     warranty_exp_date = fields.Date(string="Warranty Expiration Date")
 
+    def _get_warranty_exp_date(self, start_date):
+        if not self.product_id:
+            return False
+        duration_type = self.product_id.product_tmpl_id.warranty_type
+        duration = self.product_id.product_tmpl_id.warranty
+        if not duration_type or not duration:
+            return False
+        if duration_type == "day":
+            warranty_delta = timedelta(days=duration)
+        elif duration_type == "week":
+            warranty_delta = timedelta(weeks=duration)
+        elif duration_type == "month":
+            warranty_delta = relativedelta(months=+duration)
+        elif duration_type == "year":
+            warranty_delta = relativedelta(years=+duration)
+        else:
+            raise RuntimeError(f"Invalid warranty_type {duration_type!r}")
+        return (start_date + warranty_delta).strftime(DEFAULT_SERVER_DATE_FORMAT)
+
     @api.onchange("product_id")
     def _onchange_product_id(self):
-        self.warranty_exp_date = False
-        if (
-            self.product_id
-            and self.product_id.product_tmpl_id.warranty_type
-            and self.product_id.product_tmpl_id.warranty
-        ):
-            warranty_type = self.product_id.product_tmpl_id.warranty_type
-            time = False
-            if warranty_type == "day":
-                time = (
-                    datetime.now()
-                    + timedelta(days=self.product_id.product_tmpl_id.warranty)
-                ).strftime(DEFAULT_SERVER_DATE_FORMAT)
-            elif warranty_type == "week":
-                time = (
-                    datetime.now()
-                    + timedelta(weeks=self.product_id.product_tmpl_id.warranty)
-                ).strftime(DEFAULT_SERVER_DATE_FORMAT)
-            elif warranty_type == "month":
-                time = (
-                    datetime.now()
-                    + relativedelta(months=+self.product_id.product_tmpl_id.warranty)
-                ).strftime(DEFAULT_SERVER_DATE_FORMAT)
-            elif warranty_type == "year":
-                time = (
-                    datetime.now()
-                    + relativedelta(years=+self.product_id.product_tmpl_id.warranty)
-                ).strftime(DEFAULT_SERVER_DATE_FORMAT)
-            self.warranty_exp_date = time
+        self.warranty_exp_date = self._get_warranty_exp_date(datetime.now())
```

Comment:
IMO we can use a default value in the function for the start date (cf suggestion above).
Moreover, if the field is of Date type, we can use Date instead of Datetime.